Develop/songarden -> dev_songarden

songarden opened the pull request:
- alarm_sleep() 함수로 현재 쓰레드를 일정 시간동안 sleep_list에 보내 불필요한 CPU 점유 방지
- priority 가 높은 순으로 context switching을 하도록 변경
- CPU를 점유한 쓰레드보다 더 높은 쓰레드가 ready_list에 생기게 될 여지가 있는 모든 경우에 context_swtiching을 진행하도록 구현
- synch.c에 있는 모든 synchronization 기법들에 priority 우선순위 스케줄링 방식을 적용하도록 변경
- synchronization에 의해 block된 스레드의 우선순위를 임계영역을 소요한 스레드에게 donate하도록 구현


(MLFQS 구현부)
- load_avg scale factor를 통해 스케줄링 동적 제어 구현
- recent CPU 도입을 통해 CPU 독점에 조금 더 공평한 feedback 스케줄링 구현
- 소수점 인자인 load_avg,recent CPU 값을 계산하기 위해 고정 소수점 표현법을 활용했으며, 이를 연산하기 위해 함수로 구현

Commit:
update : not busy-waiting when sleep

Commit:
update : sleep_list sorted min to max

Commit:
update : ready_list sorted with prioirity if priority changed, Thread yield

Commit:
fix : alarm-single -> don't have to switching with awake threads

Commit:
updated : priority scheduling with synchronization

Commit:
feature : lock donating feature is added

Commit:
fix : exception : if blocked at sema , wait_to_lock will not saved

Commit:
MLFQS implement complete
1. thread_mlfqs 변수를 통해 우선순위 기부 상황을 제어
2. timer_interrupt() 함수에서 1틱마다 현재 쓰레드의 recent_cpu를 +1
3. timer_interrupt() 함수에서 1초마다 load_avg 전역 변수와 모든 스레드의 recent_cpu, priority를 재연산하도록 구현
4. 고정 소수점 연산을 고려한 함수를 사용하여 load_avg , recent_cpu , priority를 계산하도록 함

Commit:
fix : delete static var -> ready_threads

sirloinbh commented on threads/thread.c:
```diff
@@ -308,6 +316,41 @@ thread_yield (void) {
+		}
+		telem = list_next(telem);
+	}
+	// intr_set_level(old_level);
```
기존에 사용하지만 현재는 사용하지 않는 로직은 주석처리한 후 커밋하기전에 지우는 걸 추천드립니다. 어차피 git file changed에 들어가 보면 왼쪽에 다 떠요

songarden replied:
좋은 지적 감사합니다 !

sirloinbh commented on devices/timer.c:
```diff
@@ -131,8 +131,8 @@ static void
 	// printf("timer intr: %d\n", ticks);
-	time_to_wake ();
 	thread_tick ();
+	time_to_wake ();
```
c언어 뿐만 아니라 모든 프로그래밍 언어에서 함수의 호출 순서가 중요한데, 왜 두 함수들의 호출 순서를 바꾼 건지 이유가 필요해보입니다.

songarden replied:
이전에 로직 순서가 잘못되었다는 것을 이후에 깨닫고 1틱이라도 정확히 깨우고자 tick을 세고 깨울 수 있게 하려고 수정했습니다 !

Commit:
fixed: idle_thread가 ready_list에 존재하지 않도록 수정
check_running_thread() 함수에서 idle_thread가 현재 스레드라면 ready_list에 삽입하지 않고 4tick이후 자동 context switching 되도록 유도함.

sirloinbh commented on threads/thread.c:
```diff
@@ -264,13 +264,20 @@ thread_unblock (struct thread *t) {
-		thread_yield();
+		list_insert_ordered(&ready_list,&curr->elem,thread_more_priority,NULL);
+		curr->status = THREAD_READY;
+		schedule();
```
코드에 주석도 없고 커밋 메시지가 없으니 코드를 판단하기가 매우 어렵습니다. 게다가 본인이 쓴 코드도 오랜 시간이 지난 후에 보면 알아먹기 힘들어지므로 둘 중 하나는 있어야 합니다. 1. 주석 2. 커밋 메시지
옛날에는 1번이 선호됐으나 시간이 흐를 수록 2번을 선호하는 것 같습니다.

songarden replied:
필기 : commit의 습관화...

sirloinbh commented on include/threads/thread.h:
```diff
@@ -139,6 +139,12 @@ void time_to_wake (void);
 int thread_get_priority (void);
 void thread_set_priority (int);
 
+bool thread_more_priority(const struct list_elem *a_, const struct list_elem *b_,
```
어떤 함수는 thread.c의 맨 위에 정의/선언 하고, 어떤 함수는 thread.h파일에 선언하는 데 기준이 있는 지 궁금합니다.

songarden replied:
synch.c에서도 사용하는지 안하는지에 따른 차이였습니다. 원래 static으로 모두 thread.c 에 선언하려 했으나, synch에 사용하게 되어서 옮기게 되었습니다. 커밋마다 이런 사항들을 작성해 놓도록 할게요 !

sirloinbh commented on threads/synch.c:
```diff
@@ -219,13 +231,17 @@ lock_try_acquire (struct lock *lock) {
    handler. */
 void
 lock_release (struct lock *lock) {
+	struct thread *curr = thread_current();
```
이 아래에 있는 코드 변화 목록에서 *curr를 사용하는 곳이 안보이는데 선언한 이유가 있을까요?

songarden replied:
디버깅용으로 현재 스레드의 정보를 파악하려고 선언했었습니다 ! 지워야할 것 같네요

sirloinbh commented on threads/thread.c:
```diff
@@ -525,6 +531,17 @@ thread_more_priority (const struct list_elem *a_, const struct list_elem *b_,
 
+/* ready_list를 높은 우선순위 순으로 정렬하는 bool 함수 */
+bool
+thread_more_lock_priority (const struct list_elem *a_, const struct list_elem *b_,
```
코드 리뷰 예시

sirloinbh commented on threads/thread.c:
ready_threads라는 변수가 구현중에 필요해서 새로 생성했을 것 같은데 다 지운 이유가 궁금합니다.

songarden replied:
ready_list 안에 스레드들을 세는 연산이 load_avg값을 구할 때 필요한데, 이를 ready_list 출입 때 제어해보려고 했습니다.
근데 그 방법에서 생겨난 오류를 해결하지 못해서 그냥 원상태로 돌려놨습니다.

sirloinbh reviewed (commented):
나쁜 코드리뷰 예시 :
코드 잘 봤습니다 ^^
